src/CharterManager.sol

```diff
@@ -195,33 +213,49 @@ contract CharterManagerImp {
         ManagedGemJoinLike(gemJoin).exit(urp, usr, val);
     }
 
-    function frob(address gemJoin, address u, address v, address w, int256 dink, int256 dart) external allowed(u) {
-        require(u == v && w == msg.sender, "CharterManager/not-matching");
-        address urp = getOrCreateProxy(u);
+    function price(bytes32 ilk, uint256 spot) internal returns (uint256 cur) {
+        (, uint256 mat) = SpotterLike(spotter).ilks(ilk);
+        cur = rmul(rmul(spot, mat), SpotterLike(spotter).par());
+    }
 
-        bytes32 ilk = ManagedGemJoinLike(gemJoin).ilk();
-        uint256 _gate = gate[ilk];
+    function draw(bytes32 ilk, address u, address urp, address w, int256 dink, int256 dart, uint256 rate, uint256 _gate) internal {
         uint256 _nib = (_gate == 1) ? nib[ilk][u] : Nib[ilk];
+        uint256 dtab = mul(rate, uint256(dart)); // rad
+        uint256 coin = wmul(dtab, _nib);         // rad
 
-        uint256 rate;
-        if (dart > 0 && (_nib > 0 || _gate == 1)) {
-            (,rate,,,) = VatLike(vat).ilks(ilk);
+        VatLike(vat).frob(ilk, urp, urp, urp, dink, dart);
+        VatLike(vat).move(urp, w, sub(dtab, coin));
+        VatLike(vat).move(urp, vow, coin);
+    }
+
+    function validate(bytes32 ilk, address u, address urp, uint256 rate, uint256 spot, uint _gate) internal {
```

Comment:
nit: we usually are explicit with type sizes `uint _gate` should be `uint256 _gate`